TheSadRogue.Primitives/SpatialMaps/AutoSyncSpatialMap.cs

```diff
@@ -0,0 +1,622 @@
+        /// exists in the spatial map and if the new position is not already filled by some other item; false otherwise.
+        /// </summary>
+        /// <param name="item">Item to move.</param>
+        /// <param name="targetX">X-value of the location to move item to.</param>
+        /// <param name="targetY">Y-value of the location to move item to.</param>
+        /// <returns>true if the given item can be moved to the given position; false otherwise.</returns>
+        public bool CanMove(T item, int targetX, int targetY) => _spatialMap.CanMove(item, targetX, targetY);
+
+        /// <summary>
+        /// Returns true if the item at the current position specified can be moved to the target position, eg. if an item exists
+        /// at the current
+        /// position and the new position is not already filled by some other item; false otherwise.
+        /// </summary>
+        /// <param name="current">Location to move items from.</param>
+        /// <param name="target">Location to move items to.</param>
+        /// <returns>
+        /// true if all items at the position current can be moved to the position target; false if one or more items
+        /// cannot be moved.
+        /// </returns>
+        public bool CanMoveAll(Point current, Point target) => _spatialMap.CanMoveAll(current, target);
+
+        /// <summary>
+        /// Returns true if the item at the current position specified can be moved to the target position, eg. if an item exists
+        /// at the current
+        /// position and the new position is not already filled by some other item; false otherwise.
+        /// </summary>
+        /// <param name="currentX">X-value of the location to move items from.</param>
+        /// <param name="currentY">Y-value of the location to move items from.</param>
+        /// <param name="targetX">X-value of the location to move items to.</param>
+        /// <param name="targetY">Y-value of the location to move items to.</param>
+        /// <returns>
+        /// true if all items at the position current can be moved to the position target; false if one or more items
+        /// cannot be moved.
+        /// </returns>
+        public bool CanMoveAll(int currentX, int currentY, int targetX, int targetY) => _spatialMap.CanMoveAll(currentX, currentY, targetX, targetY);
+
+        /// <summary>
+        /// Moves the item specified to the position specified, updating its <see cref="IPositionable.Position"/> field
+        /// accordingly. Throws ArgumentException if the item does not exist in the spatial map or if the position is
+        /// already filled by some other item.
+        /// </summary>
+        /// <param name="item">Item to move.</param>
+        /// <param name="target">Location to move item to.</param>
+        public void Move(T item, Point target)
+        {
+            item.Position = target;
```

Comment:
Don't you need something like `_spatialMap.Move`? Also, I don't see code to throw an exception as the method summary says. These single items for `Move` are in other places. I'm assuming my knowledge about how this should work is just inaccurate. :)

Reply:
The auto-spatial map is subscribing to its item's `PositionChanging` event, and the handler calls `_spatialMap.Move`; so setting the `Position` field of the item, thus triggering that event, is really all we need to do, assuming two things:
1. The Position field is not out of sync with the spatial map's position (auto-sync map is designed to disallow this)
2. The object is actually in the spatial map.

The second assumption is not necessarily valid, admittedly.  If `Move` is called on an item which is not in the spatial map, the current logic will set its position field and return, which is not consistent with the behavior of non auto-syncing spatial maps.

I can add a check to see if the item is present; but this does come with a performance cost.

In the case of `TryMove`, I do call `TryMove` first then set the position, because I need to verify that the move will succeed before setting the position field, rather than just letting the `Move` handler fail.  I could implement `Move` via `TryMove`, but it would come with a similar performance cost.

I'll benchmark what the cost is and revisit.

Reply:
Overhead was [15-45 ns](https://github.com/thesadrogue/TheSadRogue.Primitives/files/10997202/multi_contains_map.txt).

This seems fine, especially since it doesn't affect things when you simply set via the `Position` field.  Made a change to fix this behavior so that if something is moved that isn't in the spatial map it throws the appropriate exception.